### ISSUE TITLE
Support for 'normal' SWF loading/instantiating

### ISSUE THREAD

**Issue by ajwfrost** [Corsaair/redtamarin#76]
Hi   

I was trying to embed some data in a SWF file that was loaded by the redtamarin shell and it's made everything stop working. Having done some debugging, this is because I'm getting now an abc block that has two scripts in it: (1) my main class which is what I had before adding the 'embed' statement, and (2) the generated class which contains the byte array. However, retamarin is set up normally to only execute the main (final) script - and I can see this happening, both from the method ID that's returned from `initAllScripts` and from the output when I enable verbose mode.   

So I'm planning to expand upon the support for SWF files in order to allow two things:
1.   It will parse the 'SymbolClass' block from the SWF file and will instantiate an object using the class given at symbol ID 0 (i.e. how Flash/AIR works)
2.   It will add the link between a class name and the relevant SWF 'DefineBinaryData' block so that the AS code can access the embedded byte array.   

I'm doing this off-line for now but have ordered books on git/github so will raise these as pull requests at some point should you be interested!
Thanks

**Comment by zwetan** [comment]
OK ... maybe ...

> embed some data in a SWF file

using the `[embed]` metadata is not supported

> an abc block that has two scripts

should work, more exactly many ABC blocks inside a SWF file
should be able to be parsed without issues
at the condition that only the last ABC block got the executable bit

but again you need to provide
- your whole setup to compile that problematic SWF
- you're talking about `[embed]`, we don't support that
- really we need more info so we can replicate it on our side

few things to remember
- our main compiler is ASC, not MXMLC
- redtamarin should support to load SWF  
  but not the same SWF that an MXMLC can compile  
  and that Flash Player or AIR can parse
- it is not our goal to re-implement SWF parsing  
  inside the AVM2 to support special things like `[embed]`

A SWF file is composed of many tags,
we only support the **doABC2** tag

for ex: if you embed a PNG image as a symbol in a SWF
redtamarin should be able to read the SWF but will ignore
and not process the PNG symbol

so about

> So I'm planning to expand upon the support for SWF files in order to allow two things:
> 1. It will parse the 'SymbolClass' block from the SWF file and will instantiate an object using the class given at symbol ID 0 (i.e. how Flash/AIR works)
> 2. It will add the link between a class name and the relevant SWF 'DefineBinaryData' block so that the AS code can access the embedded byte array.

Yes that would be the way to do it but that's quite advanced work
that has a lot of implications in how the avmshell load and initialize code

for example, you gonna have to inject the definition you obtain from the 'DefineBinaryData' tag
into the domain pool, so when the context switch to user code a user can do
`import path.to.definition` to be able to use it

but then the definition itself need to be supported
if the binary data is ABC code, then it is similar to load dynamically an ABC file
but if the binary data is a bitmap or a sound symbol
that will not work as we have no native definitions that support that yet

What I'm trying to say is what you are proposing is not a simple patch
so I would advise that
- before making such big commit, do a simple one
- for example: work on a simple native class, and make a pull request
- or just sending us a diff so we can patch and test the changes  
  is something that could work too
- or take an already existing issue and work on that

We welcome all contributions but we also have a stability and trust concern
now Redtamarin is used in production on servers, we can not accept blindly any patches
we need to be able to test and review those patches before integrating them

Also take into account that Gihub is a mirror of our main repository
we made it in such a way that we can take pull requests
but we could also be working on things that are not yet mirrored to github

take something like
#60 sync with avmplus from 2015-12-16
if we are working on that (and that's a huge patch)
if you're working in parallel on something as big on your side
this could create integration conflicts

so whatever you do, communicate regularly with us

**Comment by zwetan** [comment]
see [How To Contribute](https://github.com/Corsaair/redtamarin/wiki/HowToContribute)
about reporting errors and submitting patches

**Comment by ajwfrost** [comment]
Hi  
Thanks for the comments here. I'm aware of the current usage of the ASC to create .abc files rather than the focus being on SWF support (other tags etc). I guess in a way, what I'm trying to do is to make it "easier" to use redtamarin as the runtime for some non-gui, server-side AS3 software but to fit into a normal workflow used by most AS3 developers at the moment (i.e. generating a SWF file using Flash Builder or FlashDevelop etc). So adding support for other tags from the SWF is one way to achieve that.  
I take your point about the size of this and the merging/integration tasks particularly if you're doing the update from several years' worth of Adobe changes to the avmplus project! But I believe it can be pretty independently implemented...  
In terms of contributing, I guess there are three options: (a) I work on stuff according to my personal priorities and you guys integrate it into redtamarin; (b) I work on this stuff but it stays in a fork of redtamarin; or (c) I could just do this separately, starting off with (the new!) avmplus and adding to that directly. Which would mean that I don't get the benefit of all your existing stuff, but equally I don't want to start pulling this in a direction you're not comfortable with, or ending up adding some fragmentation..  
How about for now (at least for this one) I work on it in the fork, and then when you're done with the avmplus updates I could up-merge stuff on my side to make things easier for a pull request back again? Let me know what you think..  
In terms of actually making the changes, I'd already created something like this a couple of years ago when I was playing around (but the code never made it to an open source repo) so it should be okay to do :-)  
BTW - to reproduce the set-up I've got then please see the project at [https://github.com/Corsaair/redtamarin/issues/url](url) and uncomment the lines in MainApp.as.  
cheers

**Comment by zwetan** [comment]
**non-goals:**
support parsing of other SWF tags than doABC
this will not happen inside Redtamarin as this force us to parse SWF tags in C/C++
again we're not here to replicate the Flash Player

Parsing of other SWF tags can and probably will be done in an external library
at a later time, still ... this functionality is not planed to be at the core of Redtamarin API

**goals:**
Being able to run a SWF produced by an IDE like Flash Builder
this is implemented with `Program.start()`

example
```as3
package 
{
	import flash.utils.getDefinitionByName;
	
	import C.stdlib.EXIT_SUCCESS;
	import C.stdlib.exit;
	import C.unistd.getlogin;
	
	import encoding.base64.encodeBase64;
	
	import shell.Program;
	import shell.Runtime;
	
	public class test_one
	{
		
		public function test_one()
		{
			// the user code will execute there 
			trace( "hello world" );
			trace( "username: " + getlogin() );
			trace( "redtamarin  v" + Runtime.version.toString( true ) );
			
			var str:String = "hello world";
			var b64:String = encodeBase64( str );
			trace( "string: " + str );
			trace( "base64: " + b64 );
			
			try
			{
				var a:* = getDefinitionByName( "Something" );
			}
			catch( e:Error )
			{
				trace( "error" );
				trace( e );
			}
			
			exit( EXIT_SUCCESS );
		}
		
	}
	
}

import shell.Program;
Program.start( "test_one" );
```

In Flash Builder
  1. use `redtamarin.swc` instead of `playerglobal.swc`
  2. compile from the IDE
  3. then run `$ redshell bin-debug/test_one.swf`


Being able to dynamically load a SWF file at runtime
this is implemented with `Runtime.load()`

example
```as3
import shell.Runtime;

var loaded:Boolean = Runtime.load( "myfile.swf" );
trace( "loaded" + loaded );

// from that point you can use any definitions declared in myfile.swf
import test.something.MyClass; //defined in myfile.swf

var test:* = new MyClass();
```

All these are implemented in 0.4.2
